### PR TITLE
Profiling  + performance optimization

### PR DESCRIPTION
**Pull request by jakobbraun**

### CODE REVIEW

File: src/main/java/com/exasol/adapter/dynamodb/queryplanning/RemoteTableQuery.java [hunk 1]
```diff
@@ -37,14 +37,27 @@ public RemoteTableQuery(final TableMapping fromTable, final List<ColumnMapping>
         this.selectList = selectList;
         this.pushDownSelection = pushDownSelection;
         this.postSelection = postSelection;
-        this.postSelectionsColumns = new InvolvedColumnCollector().collectInvolvedColumns(getPostSelection());
+        this.requiredColumns = calculateRequiredColumns();
+
```
Comment (AnastasiiaSergienko) [inline src/main/java/com/exasol/adapter/dynamodb/queryplanning/RemoteTableQuery.java:41]: remove empty line

File: src/main/java/com/exasol/dynamodb/DynamodbConnectionFactory.java [hunk 2]
```diff
@@ -49,6 +49,7 @@ public AmazonDynamoDB getLowLevelConnection(final String uri, final String user,
         return clientBuilder.build();
     }
 
+
```
Comment (AnastasiiaSergienko) [inline src/main/java/com/exasol/dynamodb/DynamodbConnectionFactory.java:52]: ?

File: src/main/java/com/exasol/dynamodb/DynamodbValueToJsonConverter.java [hunk 3]
```diff
@@ -16,6 +16,26 @@
...
+     * Get a singleton instance of {@link DynamodbValueToJsonConverter}.
+     *
+     * @return instance of {@link DynamodbValueToJsonConverter}
+     */
+    public static DynamodbValueToJsonConverter getInstance() {
+        return INSTANCE;
+    }
+
+    // This is an performance optimization as JsonProvider.provider() is quite slow
+    private static final JsonProvider JSON = JsonProvider.provider();
```
Comment (AnastasiiaSergienko) [inline src/main/java/com/exasol/dynamodb/DynamodbValueToJsonConverter.java:38]: please move up to the other constant

File: src/test/java/com/exasol/adapter/dynamodb/AbstractExasolTestInterface.java [hunk 4]
```diff
@@ -105,14 +106,22 @@ public void createAdapterScript() throws SQLException {
 
     private void addDebuggerOptions(final StringBuilder statementBuilder) {
         final String hostIp = getTestHostIpAddress();
+        final StringBuilder jvmOptions = new StringBuilder();
         if (hostIp != null) {
-            final StringBuilder jvmOptions = new StringBuilder("-javaagent:/buckets/bfsdefault/default/"
-                    + JACOCO_JAR_NAME + "=output=tcpclient,address=" + hostIp + ",port=3002");
+            jvmOptions
+                    .append("-javaagent:/buckets/bfsdefault/default/" + JACOCO_JAR_NAME + "=output=tcpclient,address=")
```
Comment (AnastasiiaSergienko) [inline src/test/java/com/exasol/adapter/dynamodb/AbstractExasolTestInterface.java:112]: Don't use + operators inside append method

File: src/test/java/com/exasol/adapter/dynamodb/DynamodbAdapterTestAwsIT.java [hunk 5]
```diff
@@ -57,10 +57,29 @@ static void beforeAll() throws SQLException, BucketAccessException, InterruptedE
     @Test
     void testCountAllRows() throws SQLException {
         final ResultSet resultSet = exasolTestInterface.getStatement()
-                .executeQuery("SELECT COUNT(\"KEY\") FROM OPENLIBRARY");
+                .executeQuery("SELECT \"KEY\", \"REVISION\" FROM OPENLIBRARY");
+        resultSet.next();
```
Comment (AnastasiiaSergienko) [inline src/test/java/com/exasol/adapter/dynamodb/DynamodbAdapterTestAwsIT.java:61]: What does it test?

File: src/test/java/com/exasol/adapter/dynamodb/DynamodbAdapterTestAwsIT.java [hunk 6]
```diff
@@ -76,9 +95,9 @@ void testSelectSingleRow() throws SQLException {
     @Test
     void testCountAuthorsTable() throws SQLException {
         final ResultSet resultSet = exasolTestInterface.getStatement()
-                .executeQuery("SELECT COUNT(\"KEY\") FROM OPENLIBRARY_AUTHORS");
+                .executeQuery("SELECT \"KEY\" FROM OPENLIBRARY_AUTHORS");
         resultSet.next();
-        final int count = resultSet.getInt(1);
-        assertThat(count, equalTo(858180));
+        // final int count = resultSet.getInt(1);
```
Comment (AnastasiiaSergienko) [inline src/test/java/com/exasol/adapter/dynamodb/DynamodbAdapterTestAwsIT.java:100]: I think it's better to remove or uncomment before merging

File: src/main/java/com/exasol/adapter/dynamodb/documentpath/LinearDocumentPathWalker.java [hunk 7]
```diff
@@ -27,8 +27,8 @@ public LinearDocumentPathWalker(final DocumentPathExpression pathExpression) {
      * {@link DocumentPathWalker#walkThroughDocument(DocumentNode)}, this method returns one single value.
      *
      * @param rootNode document to walk through
-     * @return documents attribute described in {@link DocumentPathExpression}
-     * @throws DocumentPathWalkerException if defined path does not exist in the given document
+     * @return documents attribute described in {@link DocumentPathExpression} or null if the path does not exist in the
```
[the remark this answers is not in the document]
Comment (jakobbraun) [reply src/main/java/com/exasol/adapter/dynamodb/documentpath/LinearDocumentPathWalker.java:30]: The exception is thrown in the constructor. There the description fits.

File: src/test/java/com/exasol/adapter/dynamodb/DynamodbAdapterTestAwsIT.java (hunk 5 above)
Comment (jakobbraun) [reply src/test/java/com/exasol/adapter/dynamodb/DynamodbAdapterTestAwsIT.java:61]: This are performance tests. I only use them to measure the runtime

**Pull request merged**